File: HexaFlipTests/Source/Core/JCSFlipMoveTest.m
//
//  JCSFlipMoveTest.m
//  HexaFlip
//
//  Created by Christian Schuster on 18.07.12.
//  Copyright (c) 2012 Christian Schuster. All rights reserved.
//

#import "JCSFlipMove.h"
#import "JCSFlipMoveInputDelegate.h"

#import "OCMock.h"

#import <objc/runtime.h>

// category with methods to track blocks that are scheduled during move input
@implementation JCSFlipMove (BlockCollectingTest)

// dictionary for collecting and sorting blocks
// key: block (dispatch_block_t), value: delay (NSNumber *)
static NSMutableDictionary *_delayForBlock;

- (void)clearCollectedBlocks {
    _delayForBlock = [NSMutableDictionary dictionary];
}

// will be swizzled with invokeOnMainQueueAfterDelay:block:
- (void)collectWithDelay:(double)seconds block:(dispatch_block_t)block {
    [_delayForBlock setObject:@(seconds) forKey:block];
}

- (void)invokeCollectedBlocks {
    NSArray *blocksSortedByDelay = [_delayForBlock keysSortedByValueUsingComparator:^NSComparisonResult(NSNumber *delay1, NSNumber *delay2) {
        return [delay1 compare:delay2];
    }];
    for (dispatch_block_t block in blocksSortedByDelay) {
        block();
    }
}

@end

@interface  JCSFlipMove ()

// make selector visible to test
- (void)dispatchToMainQueueAfterDelay:(double)seconds block:(dispatch_block_t)block;

@end

@interface JCSFlipMoveTest : XCTestCase
@end

@implementation JCSFlipMoveTest {
    // each entry is an NSArray containing a NSNumber (delay, index 0) and a dispatch_block_t (block, index 1)
    NSMutableArray *_blockInfo;
}

- (void)testInitMove {
    JCSFlipMove *move = [[JCSFlipMove alloc] initWithStartRow:1 startColumn:4 direction:JCSHexDirectionNE];
    
    XCTAssertFalse(move.skip);
    XCTAssertEqual(move.startRow, 1);
    XCTAssertEqual(move.startColumn, 4);
    XCTAssertEqual(move.direction, JCSHexDirectionNE);
}

- (void)testInitSkip {
    JCSFlipMove *move = [[JCSFlipMove alloc] init];
    
    XCTAssertTrue(move.skip);
    XCTAssertThrows([move startRow]);
    XCTAssertThrows([move startColumn]);
    XCTAssertThrows([move direction]);
}

- (void)testMutable {
    JCSFlipMove *move = [[JCSFlipMove alloc] init];
    
    move.skip = NO;
    move.startRow = 1;
    move.startColumn = 2;
    move.direction = JCSHexDirectionNW;
    
    XCTAssertFalse(move.skip);
    XCTAssertEqual(move.startRow, 1);
    XCTAssertEqual(move.startColumn, 2);
    XCTAssertEqual(move.direction, JCSHexDirectionNW);
}

- (void)testCopy {
    JCSFlipMove *move = [[JCSFlipMove alloc] initWithStartRow:32 startColumn:12 direction:JCSHexDirectionE];
    JCSFlipMove *copy = [move copy];
    
    XCTAssertFalse(move == copy);
    XCTAssertTrue([copy class] == [JCSFlipMove class]);
    XCTAssertEqual(move.skip, copy.skip);
    XCTAssertEqual(move.startRow, copy.startRow);
    XCTAssertEqual(move.startColumn, copy.startColumn);
    XCTAssertEqual(move.direction, copy.direction);
}

- (void)doTestPerformInputWithMove:(JCSFlipMove *)move {
    OCMockObject<JCSFlipMoveInputDelegate> *delegateMock = [OCMockObject mockForProtocol:@protocol(JCSFlipMoveInputDelegate)];
    [delegateMock setExpectationOrderMatters:YES];
    if (!move.skip) {
        [[[delegateMock expect] andReturnValue:@(YES)] inputSelectedStartRow:move.startRow startColumn:move.startColumn];
        [[delegateMock expect] inputSelectedDirection:move.direction startRow:move.startRow startColumn:move.startColumn];
        [[delegateMock expect] inputClearedDirection:move.direction startRow:move.startRow startColumn:move.startColumn];
        [[delegateMock expect] inputClearedStartRow:move.startRow startColumn:move.startColumn];
    }
    [[delegateMock expect] inputConfirmedWithMove:move];
    
    // swizzle delayed invocation to cut down test time and avoid using the main queue
    Method original = class_getInstanceMethod(move.class, @selector(dispatchToMainQueueAfterDelay:block:));
    Method swizzled = class_getInstanceMethod(move.class, @selector(collectWithDelay:block:));
    method_exchangeImplementations(original, swizzled);
    @try {
        // clear collected blocks
        [move clearCollectedBlocks];
        
        // invoke the method to be tested, collecting any blocks that are scheduled
        [move performInputWithMoveInputDelegate:delegateMock];
        
        // invoke collected blocks in correct order
        [move invokeCollectedBlocks];
        
        [delegateMock verify];
    } @finally {
        // swizzle back
        method_exchangeImplementations(original, swizzled);
    }
}

- (void)testPerformInputWhenValidNormalMove {
    [self doTestPerformInputWithMove:[[JCSFlipMove alloc] initWithStartRow:17 startColumn:23 direction:JCSHexDirectionNW]];
}

- (void)testPerformInputWhenValidSkipMove {
    [self doTestPerformInputWithMove:[[JCSFlipMove alloc] init]];
}

@end

File: HexaFlipTests/Source/Core/JCSFlipMoveTest.swift
//
//  JCSFlipMoveTest.swift
//  HexaFlip
//
//  Created by Christian Schuster on 14.09.14.
//  Copyright (c) 2014 Christian Schuster. All rights reserved.
//

import Foundation
import XCTest

class JCSFlipMoveTest : XCTestCase {
    
    func testInitNormalMove() {
        let move = JCSFlipMove(startRow: 1, startColumn: 4, direction: JCSHexDirection.NE)
        
        XCTAssertFalse(move.skip)
        XCTAssertEqual(move.startRow, 1)
        XCTAssertEqual(move.startColumn, 4)
        XCTAssertEqual(move.direction, JCSHexDirection.NE)
    }
    
    func testInitSkipMove() {
        let move = JCSFlipMove()
        
        XCTAssertTrue(move.skip)
    }
    
    func testMutable() {
        let move = JCSFlipMove()
        
        move.skip = false
        move.startRow = 1
        move.startColumn = 2
        move.direction = JCSHexDirection.NW
        
        XCTAssertFalse(move.skip)
        XCTAssertEqual(move.startRow, 1)
        XCTAssertEqual(move.startColumn, 2)
        XCTAssertEqual(move.direction, JCSHexDirection.NW)
    }
    
    func testCopy() {
        let move = JCSFlipMove(startRow: 32, startColumn: 12, direction:JCSHexDirection.E)
        
        let copy = move.copy() as JCSFlipMove
        
        XCTAssertTrue(move !== copy)
        XCTAssertTrue(move.isKindOfClass(JCSFlipMove))
        XCTAssertEqual(move.skip, copy.skip)
        XCTAssertEqual(move.startRow, copy.startRow)
        XCTAssertEqual(move.startColumn, copy.startColumn)
        XCTAssertEqual(move.direction, copy.direction)
    }
    
    func testPerformInputWhenNormalMove() {
        let move = JCSFlipMove(startRow: 17, startColumn: 23, direction: JCSHexDirection.NW)
        let delegate = MockDelegate(expectedMove: move, expectation: expectationWithDescription("invocation sequence"))
        
        move.performInputWithMoveInputDelegate(delegate)
        
        waitForExpectationsWithTimeout(0.55, nil)
    }
    
    func testPerformInputWhenSkipMove() {
        let move = JCSFlipMove()
        let delegate = MockDelegate(expectedMove: move, expectation: expectationWithDescription("invocation sequence"))
        
        move.performInputWithMoveInputDelegate(delegate)
        
        waitForExpectationsWithTimeout(0.05, nil)
    }
    
    func testCompareToMoveByValueAscending() {
        let moveA = JCSFlipMove()
        moveA.value = 0.0
        let moveB = JCSFlipMove()
        moveB.value = 1.0
        
        XCTAssertEqual(moveA.compareByValueTo(moveB), NSComparisonResult.OrderedAscending, "expected move A < move B")
    }
    
    func testCompareToMoveByValueSame() {
        let moveA = JCSFlipMove()
        moveA.value = 1.23
        let moveB = JCSFlipMove()
        moveB.value = 1.23
        
        XCTAssertEqual(moveA.compareByValueTo(moveB), NSComparisonResult.OrderedSame, "expected move A = move B")
    }
    
    func testCompareToMoveByValueDescending() {
        let moveA = JCSFlipMove()
        moveA.value = 0.3
        let moveB = JCSFlipMove()
        moveB.value = -0.1
        
        XCTAssertEqual(moveA.compareByValueTo(moveB), NSComparisonResult.OrderedDescending, "expected move A > move B")
    }
    
    private class MockDelegate : NSObject, JCSFlipMoveInputDelegate {
        
        private enum Invocation : String {
            case SelectStartPoint = "select start point"
            case SelectDirection = "select direction"
            case ClearDirection = "clear direction"
            case ClearStartPoint = "clear start point"
            case Confirm = "confirm"
            case None = "none"
        }
        
        let expectedMove: JCSFlipMove
        let expectation: XCTestExpectation
        var nextInvocation: Invocation
        
        init(expectedMove: JCSFlipMove, expectation: XCTestExpectation) {
            self.expectedMove = expectedMove
            self.expectation = expectation
            self.nextInvocation = (expectedMove.skip ? Invocation.Confirm : Invocation.SelectStartPoint)
        }
        
        func inputSelectedStartRow(startRow: Int, startColumn: Int) -> Bool {
            XCTAssertEqual(nextInvocation, Invocation.SelectStartPoint, "unexpected invocation")
            XCTAssertEqual(startRow, expectedMove.startRow, "start row")
            XCTAssertEqual(startColumn, expectedMove.startColumn, "start column")
            nextInvocation = Invocation.SelectDirection
            return true
        }
        
        func inputClearedStartRow(startRow: Int, startColumn: Int) {
            XCTAssertEqual(nextInvocation, Invocation.ClearStartPoint, "unexpected invocation")
            XCTAssertEqual(startRow, expectedMove.startRow, "start row")
            XCTAssertEqual(startColumn, expectedMove.startColumn, "start column")
            nextInvocation = Invocation.Confirm
        }
        
        func inputModifiedStartRow(startRow: Int, startColumn: Int, previousStartRow: Int, previousStartColumn: Int) -> Bool {
            XCTFail("unexpected invocation")
            return true
        }
        
        func inputSelectedDirection(direction: JCSHexDirection, startRow: Int, startColumn:Int) {
            XCTAssertEqual(nextInvocation, Invocation.SelectDirection, "unexpected invocation")
            XCTAssertEqual(direction, expectedMove.direction, "direction")
            XCTAssertEqual(startRow, expectedMove.startRow, "start row")
            XCTAssertEqual(startColumn, expectedMove.startColumn, "start column")
            nextInvocation = Invocation.ClearDirection
        }
        
        func inputClearedDirection(direction: JCSHexDirection, startRow: Int, startColumn:Int) {
            XCTAssertEqual(nextInvocation, Invocation.ClearDirection, "unexpected invocation")
            XCTAssertEqual(direction, expectedMove.direction, "direction")
            XCTAssertEqual(startRow, expectedMove.startRow, "start row")
            XCTAssertEqual(startColumn, expectedMove.startColumn, "start column")
            nextInvocation = Invocation.ClearStartPoint
        }
        
        func inputCancelled() {
            XCTFail("unexpected invocation")
        }
        
        func inputConfirmedWithMove(move: JCSFlipMove) {
            XCTAssertEqual(nextInvocation, Invocation.Confirm, "unexpected invocation")
            XCTAssertEqual(move, expectedMove, "move")
            nextInvocation = Invocation.None
            expectation.fulfill()
        }
    }
}
